reduce/deprecate the usage of the ambiguous term "testnet"

q9f opened the pull request:
~~This pull request discourages users to use the Ropsten testnet by changing the default `--testnet` to the more stable and less populated Görli testnet.~~ The `--testnet` flag is now marked as deprecated but still points to Ropsten.

In general, this pull request tries to reduce or completely eliminate the term `testnet` due to it's ambiguity. Internally, `ropsten` will be used instead wherever possible.

Tests, wherever possible, now use `Ropsten` configs instead of `Testnet`.

The only exception, however, this keeps the Ropsten database in the `testnet/` data-dir to avoid unecessary database migrations.

~~This is a breaking change and should be considered carefully.~~ Nothing should break now.

karalabe commented:
I'm sorry, but this will never happen. If Ropsten ever dies, we can remove `--testnet` and then just keep named networks from that point onward. We however cannot just silently change networks. This is an API surface that has been stable for 5 years and we're not about to break it.

karalabe commented:
All the code changes that do the internal renames from Testnet to Ropsten are completely valid and very good changes, so that's definitely welcome if you want to pursue that.

The only change I don't think we can accept is rewiring `--testnet` to goerli from ropsten.

q9f commented:
> this will never happen

:cry: Ok, is that final?

I gave a talk at EthCC-3 to emphasize my motivation on why we should change defaults.

[![Screenshot at 2020-04-01 17-35-06](https://user-images.githubusercontent.com/58883403/78156581-45247400-743f-11ea-848f-07e2d5219e84.png)](https://youtu.be/uzGEmO7H5aU?t=296)

https://youtu.be/uzGEmO7H5aU?t=296

ligi commented:
It might/will break things and I agree it is not a good thing to just change the behavior of a flag. Would just deprecating this flag work for you? With a big warning and maybe even a time penalty. IMHO this is the best thing. There is no more testnet singleton. There are multiple testnets ..

q9f commented:
> Would just deprecating this flag work for you?

Yes, that's an idea.

q9f commented:
Thanks for your feedback. I restored the `--testnet` flag to point to Ropsten and added a deprecation warning.

karalabe commented on README.md:
```diff
@@ -81,23 +81,24 @@ network, you want to join the **test** network with your node, which is fully eq
+   states.
  * Instead of using the default data directory (`~/.ethereum` on Linux for example), `geth`
-   will nest itself one level deeper into a `testnet` subfolder (`~/.ethereum/testnet` on
+   will nest itself one level deeper into a `goerli/` subfolder (`~/.ethereum/goerli` on
```
I think you accidentally added a slash after `goerli/`. If deliberate, could you expand why?

q9f replied:
I always do this to indicate sub-directories. But I can remove it.

karalabe commented on README.md:
```diff
@@ -107,17 +108,24 @@ accounts available between them.*
 
+### Full node on the Ropsten test network
+
+In addition to Görli and Rinkeby, Go-Ethereum also supports the ancient Ropsten testnet.
```
A general naming convention: `go-ethereum` (all lowercase) is the repo/project and `Geth` (capitalized) is the thing that people run. In this particular case you can use either, but if you choose the former, please keep it lowercase and ideally put it between ticks.

A bit of a special flake is if you want to start a sentence with `go-ethereum`, because then lowercase looks weird. In those cases we do use `Go Ethereum` capitalized without the dash, but generally try to keep those constructs to the minimum.

karalabe commented on README.md:
````diff
@@ -107,17 +108,24 @@ accounts available between them.*
+
+```shell
+$ geth --ropsten console
+```
````
Perhaps also add a note that for legacy reasons, this uses the `testnet` folder on disk, not the `ropsten`. Alternatively, we need a migration path (if the old `testnet` path exists, use that, otherwise use `ropsten`).

karalabe commented on cmd/geth/main.go:
```diff
@@ -261,12 +262,28 @@ func prepare(ctx *cli.Context) {
+		log.Info("Starting Geth in ephemeral dev mode...")
+	} else if !ctx.GlobalIsSet(utils.NetworkIdFlag.Name) {
+		log.Info("Starting Geth on Ethereum mainnet...")
+	}
```
When doing more than 1 `else` branch, please use a `switch` statement:

```go
switch {
case ctx.GlobalIsSet(utils.LegacyTestnetFlag.Name):
	log.Info("Starting Geth on Ropsten testnet...")
	log.Warn("The --testnet flag is ambiguous! Please specify one of --goerli, --rinkeby, or --ropsten.")
	log.Warn("The generic --testnet flag is deprecated and will be removed in the future!")

case  ctx.GlobalIsSet(utils.RopstenFlag.Name):
	log.Info("Starting Geth on Ropsten testnet...")

case  ctx.GlobalIsSet(utils.RinkebyFlag.Name):
	log.Info("Starting Geth on Rinkeby testnet...")

case  ctx.GlobalIsSet(utils.GoerliFlag.Name):
	log.Info("Starting Geth on Görli testnet...")

case  ctx.GlobalIsSet(utils.DeveloperFlag.Name):
	log.Info("Starting Geth in ephemeral dev mode...")

case  !ctx.GlobalIsSet(utils.NetworkIdFlag.Name):
	log.Info("Starting Geth on Ethereum mainnet...")
}
```

Also since there is no "natural" separator between the lines here (i.e. no `{}` inside the cases), you might want to leave an empty line between the case statements. Makes reading the code easier on the eye.

karalabe commented on cmd/utils/flags.go:
```diff
@@ -162,20 +162,24 @@ var (
 	NetworkIdFlag = cli.Uint64Flag{
 		Name:  "networkid",
-		Usage: "Network identifier (integer, 1=Frontier, 2=Morden (disused), 3=Ropsten, 4=Rinkeby)",
+		Usage: "Network identifier (integer, 1=Frontier, 2=Morden (disused), 3=Ropsten, 4=Rinkeby, 5=Görli)",
```
I think we can also drop `Morden` from this list. Honestly nobody will even know any more what that was.

q9f replied:
Ha, true! https://twitter.com/hudsonjameson/status/1205912735734784001

karalabe commented on cmd/utils/flags.go:
```diff
@@ -162,20 +162,24 @@ var (
 		Value: eth.DefaultConfig.NetworkId,
 	}
-	TestnetFlag = cli.BoolFlag{
+	LegacyTestnetFlag = cli.BoolFlag{ // Deprecated in favor of --goerli, --rinkeby, or --ropsten
```
Instead of adding `// Deprecated [...]`, which is redundant with the flag comment and has no special meaning, you could add a `// TODO(q9f): Remove after Ropsten is discontinued`. That leaves a clear mark in the code that something funky is going on here, there's a hint as to what it is and there's also a username whom to contact for more details.

karalabe commented on cmd/utils/flags.go:
```diff
@@ -774,10 +778,10 @@ var (
 // MakeDataDir retrieves the currently requested data directory, terminating
 // if none (or the empty string) is specified. If the node is starting a testnet,
-// the a subdirectory of the specified datadir will be used.
+// then a subdirectory of the specified data-dir will be used.
```
We use `datadir` throughout the code, we never use `data-dir`. Lets not introduce a new notation. Feel free to choose between `datadir` or `data directory` :)

karalabe commented on core/genesis.go:
```diff
@@ -246,8 +246,12 @@ func (g *Genesis) configOrDefault(ghash common.Hash) *params.ChainConfig {
+	case ghash == params.RinkebyGenesisHash:
+		return params.RinkebyChainConfig
+	case ghash == params.GoerliGenesisHash:
+		return params.GoerliChainConfig
```
This code path is strange. Everything worked until now without having explicit clauses for Rinkeby and co. Maybe this is dead code? Perhaps we should investigate it a bit.

q9f replied:
I'm not entirely sure. It's used to setup genesis block in various places and detects known presets based on known genesis hashes.

q9f commented:
Thanks for the review, I addressed all comments but one.

Now, `MakeDataDir()` will create a `ropsten` directory if `--testnet` or `--ropsten` is specified.

However, `setDataDir()` checks of `testnet` exists and keeps using this, otherwise it uses `ropsten`. Same is true for the console command `path` in `remoteConsole()`.

q9f commented:
Friendly Monday morning bump :coffee: :green_salad:

karalabe commented:
:coffee:

karalabe commented:
Found a tiny issue. If the `testnet` folder exists, and I start Geth with `geth --testnet console`, the console history is placed in a newly created `ropsten` folder.

karalabe commented on cmd/utils/flags.go:
```diff
@@ -774,11 +777,11 @@ var (
-		if ctx.GlobalBool(TestnetFlag.Name) {
-			return filepath.Join(path, "testnet")
+		if ctx.GlobalBool(LegacyTestnetFlag.Name) || ctx.GlobalBool(RopstenFlag.Name) {
+			return filepath.Join(path, "ropsten")
```
Plopp. This is where you forgot to add the legacy `testnet` folder check.

q9f replied:
Ok. I was assuming that `MakeDataDir()` is only called when _creating_ a new datadir? we don't want to create new `testnet` subdirs, do we? Is there any instance where this is called for an existing datadir?

karalabe replied:
When you create a `console.Config`, this method is used. That will make the console believe it's in a different folder than the rest of Geth.

karalabe replied:
https://github.com/ethereum/go-ethereum/search?q=%22console.Config%22&unscoped_q=%22console.Config%22 

3 instances in this file

karalabe replied:
Btw, it does not create a datadir. The name is a bit misleading. All the methods that take the flags and "Make" some internal config out of them are called "MakeXXX". I guess this is more meaningful for "MakeEthConfig", but it just got used as a general style for all.

q9f commented:
Thanks for the careful review. I addressed the testnet subdir check in makedatadir.

q9f commented:
Not sure why the test on arm64 fails, hiccup?

```
--- FAIL: TestBalanceAt (0.22s)
    --- FAIL: TestBalanceAt/future_block (0.10s)
        ethclient_test.go:296: BalanceAt(71562b71999873db5b286df957af199ec94617f7, 1000000000) error = "context deadline exceeded", want "header not found"
FAIL
coverage: 14.0% of statements
FAIL	github.com/ethereum/go-ethereum/ethclient	3.392s
```

karalabe commented on cmd/geth/chaincmd.go:
```diff
@@ -194,7 +195,7 @@ Use "ethereum dump 0" to dump the genesis block.`,
 			utils.AncientFlag,
 			utils.CacheFlag,
-			utils.TestnetFlag,
+			utils.RopstenFlag,
```
Please add 

```diff
diff --git a/cmd/geth/chaincmd.go b/cmd/geth/chaincmd.go
index 0a1d2e998..0351afa58 100644
--- a/cmd/geth/chaincmd.go
+++ b/cmd/geth/chaincmd.go
@@ -198,6 +198,7 @@ Use "ethereum dump 0" to dump the genesis block.`,
                        utils.RopstenFlag,
                        utils.RinkebyFlag,
                        utils.GoerliFlag,
+                       utils.LegacyTestnetFlag,
                        utils.SyncModeFlag,
                },
                Category: "BLOCKCHAIN COMMANDS",
```

I did it manually and wanted to push my changes to your branch, but alas you forbid it (strange).

```
$ git push afri q9-config-testnet 
ERROR: Permission to byz-f/go-ethereum.git denied to karalabe.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

q9f commented:
Thanks, I added your patch. 

I'm not forbidding you to write to my branch but apparently, the Github Web-UI removed that checkbox to allow maintainers access. I don't know why. You are not the first complaining.

karalabe commented:
No worries, was just weird because we always just push tiny nit fixes back onto PRs because it's simpler and less effort for everyone involved. PR LGTM, thanks!
